app/src/ui/diff/text-diff.tsx

```diff
@@ -1381,10 +1390,88 @@ export class TextDiff extends React.Component<ITextDiffProps, ITextDiffState> {
 
   public componentWillUnmount() {
     this.cancelSelection()
+    this.unmountWhitespaceHint()
     this.codeMirror = null
     document.removeEventListener('find-text', this.onFindText)
   }
 
+  private mountWhitespaceHint(index: number) {
+    this.unmountWhitespaceHint()
+
+    // Since we're in a bit of a weird state here where CodeMirror is mounted
+    // through React and we're in turn mounting a React component from a
+    // DOM event we want to make sure we're not mounting the Popover
+    // synchronously. Doing so will cause the popover to receiving the bubbling
+    // mousedown event (on document) which caused it to be mounted in the first
+    // place and it will then close itself thinking that it's seen a mousedown
+    // event outside of its container.
+    this.whitespaceHintMountId = requestAnimationFrame(() => {
+      this.whitespaceHintMountId = null
+      const cm = this.codeMirror
+
+      if (cm === null) {
+        return
+      }
+
+      const container = document.createElement('div')
+      container.style.position = 'absolute'
+      const scroller = cm.getScrollerElement()
+
+      const diffSize = getLineWidthFromDigitCount(
+        getNumberOfDigits(this.state.diff.maxLineNumber)
+      )
+
+      const lineY = cm.heightAtLine(index, 'local')
+      // We're positioning relative to the scroll container, not the
+      // sizer or lines so we'll have to account for the gutter width and
+      // the hunk handle.
+      const style: React.CSSProperties = { left: diffSize * 2 + 10 }
+      let caretPosition = PopoverCaretPosition.LeftTop
+
+      // Offset down by 10px to align the popover arrow.
+      container.style.top = `${lineY - 10}px`
+
+      // If the line is further than 50% down the viewport we'll flip the
+      // popover to point upwards so as to not get hidden beneath (or above)
+      // the scroll boundary.
+      if (lineY - scroller.scrollTop > scroller.clientHeight / 2) {
+        caretPosition = PopoverCaretPosition.LeftBottom
+        style.bottom = -35
+      }
+
+      scroller.appendChild(container)
+      this.whitespaceHintContainer = container
+
+      ReactDOM.render(
+        <WhitespaceHintPopover
+          caretPosition={caretPosition}
+          onDismissed={this.unmountWhitespaceHint}
+          onHideWhitespaceInDiffChanged={
+            this.props.onHideWhitespaceInDiffChanged
+          }
+          style={style}
+        />,
+        container
+      )
+    })
+  }
+
+  private unmountWhitespaceHint = () => {
+    if (this.whitespaceHintMountId !== null) {
+      cancelAnimationFrame(this.whitespaceHintMountId)
+      this.whitespaceHintMountId = null
+    }
+
+    if (this.whitespaceHintContainer !== null) {
+      ReactDOM.unmountComponentAtNode(this.whitespaceHintContainer)
+    }
+
+    if (this.whitespaceHintContainer !== null) {
+      this.whitespaceHintContainer.remove()
+      this.whitespaceHintContainer = null
+    }
```

Comment:
Is there a reason why this is in 2 different `if`? 🤔 
```suggestion
    if (this.whitespaceHintContainer !== null) {
      ReactDOM.unmountComponentAtNode(this.whitespaceHintContainer)
      this.whitespaceHintContainer.remove()
      this.whitespaceHintContainer = null
    }
```

Reply:
> Is there a reason why this is in 2 different `if`?

Yes! And the fault is mine, I should have documented it 🤦 So `unmountComponentAtNode` can actually (or will even) end up calling `unmountWhitespaceHint` since the `componentDidUnmount` callback in `Popover` will be invoked and in turn call `onDismissed`. So depending on who ended up calling `unmountWhitespaceHint` this can end up being reentrant which means that `this.whitespaceHintContainer` will be `null` the `unmountComponentAtNode` call.

I'll add some comments here to explain this weirdness

Reply:
https://github.com/desktop/desktop/pull/12979/commits/6e6a1f0aaf443b10d505a57bd7b2981b400484fa